solidity/contracts/Coordinator.sol

```diff
@@ -3,10 +3,36 @@ pragma solidity ^0.4.24;
 // Coordinator handles oracle service aggreements between one or more oracles.
 contract Coordinator {
 
+  struct ServiceAgreement {
+    uint256 payment;
+    uint256 expiration;
+    bytes32 jobSpecID;
+  }
+
+  mapping(bytes32 => ServiceAgreement) internal serviceAgreements;
+
   function getId(uint256 _payment, uint256 _expiration, bytes32 _jobSpecId)
     public returns (bytes32)
   {
     return keccak256(_payment, _expiration, _jobSpecId);
   }
 
+  function initiateServiceAgreement(uint256 _payment, uint256 _expiration, bytes32 _jobSpecId)
+    public
+  {
+    bytes32 id = getId(_payment, _expiration, _jobSpecId);
+
+    serviceAgreements[id] = ServiceAgreement(
+      _payment,
+      _expiration,
+      _jobSpecId
+    );
+  }
+
+  function getServiceAgreement(bytes32 _id)
+    public returns (bytes memory)
+  {
+    ServiceAgreement sa = serviceAgreements[_id];
+    return abi.encode(sa.payment, sa.expiration, sa.jobSpecID);
```

Comment:
Does this need CBOR encoding?

Reply:
I don't believe so since we're not transferring the payload in this contract.

Reply:
I think it'd be better not to ABI encode this, instead returning either the full struct itself or each of the fields. ABI encoding it will make it harder for other contracts to use, but either way is pretty much equivalent for off-chain viewers.

Reply:
@se3000 we would prefer to return the struct itself. Do you have an example of how to test that in truffle?

From what I could gather, structs only worked from within the contract. Truffle was throwing a tuple error and it seems like changes were only [recently merged](https://github.com/ethereum/web3.js/issues/1241#issuecomment-411774018) to web3.js which will now allow truffle to add support.

Reply:
I don't know off hand how to do that. I think you could use [call](https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethcall).